dds/DCPS/XTypes/TypeLookupService.cpp

```diff
@@ -436,6 +436,426 @@ bool TypeLookupService::complete_to_minimal_type_object(const TypeObject& cto, T
+    TypeDescriptor* td(new TypeDescriptor);
+    dt->descriptor_ = td;
+    DynamicTypeMap dt_map;
+    complete_to_dynamic(dt, cto, dt_map);
+  }
+
+void TypeLookupService::complete_to_dynamic(DynamicType_rch& dt,
+  const CompleteTypeObject& cto, DynamicTypeMap& dt_map)
+{
+  switch (cto.kind) {
+  // Constructed/Named types
+  case TK_ALIAS:
+    dt->descriptor_->kind = TK_ALIAS;
+    dt->descriptor_->name = cto.alias_type.header.detail.type_name;
+    dt->descriptor_->bound.length(0);
+    type_identifier_to_dynamic(dt->descriptor_->base_type, cto.alias_type.body.common.related_type, dt_map);
+    dt->member_by_index = dt->descriptor_->base_type->member_by_index;
+    break;
+  // Enumerated TKs
+  case TK_ENUM:
+    dt->descriptor_->kind = TK_ENUM;
+    dt->descriptor_->name = cto.enumerated_type.header.detail.type_name;
+    dt->descriptor_->bound.length(0);
+    for (ulong i = 0; i < cto.enumerated_type.literal_seq.length(); ++i) {
+      DynamicTypeMember_rch dtm(new DynamicTypeMember, OpenDDS::DCPS::keep_count());
+      MemberDescriptor* md(new MemberDescriptor);
+      dtm->descriptor_ = md;
+      dtm->descriptor_->name = cto.enumerated_type.literal_seq[i].detail.name;
+      dtm->descriptor_->type = dt;
+      dtm->descriptor_->is_default_label = (cto.enumerated_type.literal_seq[i].common.flags & (1 << 6));
+      dtm->descriptor_->index = i;
+      insert_dynamic_member(dt, dtm);
+    }
+    break;
+  case TK_BITMASK: {
+    dt->descriptor_->kind = TK_BITMASK;
+    dt->descriptor_->name = cto.bitmask_type.header.detail.type_name;
+    dt->descriptor_->bound.length(1);
+    dt->descriptor_->bound[0] = cto.bitmask_type.header.common.bit_bound;
+    type_identifier_to_dynamic(dt->descriptor_->element_type, TypeIdentifier(TK_BOOLEAN), dt_map);
+    for (ulong i = 0; i < cto.bitmask_type.flag_seq.length(); ++i) {
+      DynamicTypeMember_rch dtm(new DynamicTypeMember, OpenDDS::DCPS::keep_count());
+      MemberDescriptor* md(new MemberDescriptor);
+      dtm->descriptor_ = md;
+      dtm->descriptor_->name = cto.bitmask_type.flag_seq[i].detail.name;
+      type_identifier_to_dynamic(dtm->descriptor_->type, TypeIdentifier(TK_BOOLEAN), dt_map);
+      dtm->descriptor_->index = i;
+      insert_dynamic_member(dt, dtm);
+    }
+  }
+  break;
+  // Structured TKs
+  case TK_ANNOTATION:
+    dt->descriptor_->kind = TK_ANNOTATION;
+    dt->descriptor_->name = cto.annotation_type.header.annotation_name;
+    dt->descriptor_->bound.length(0);
+    for (ulong i = 0; i < cto.annotation_type.member_seq.length(); ++i) {
+      DynamicTypeMember_rch dtm(new DynamicTypeMember, OpenDDS::DCPS::keep_count());
+      MemberDescriptor* md(new MemberDescriptor);
+      dtm->descriptor_ = md;
+      complete_annotation_member_to_member_descriptor(dtm->descriptor_, cto.annotation_type.member_seq[i], dt_map);
+      dtm->descriptor_->index = i;
+      dtm->descriptor_->id = i; //Clayton: I'm not entirely sure what this should be for annotation
```

Comment:
Members of annotations don't really have ids like members of structs and unions, but the definition for this particular property 7.5.2.7.5 specifies it's for an aggregated types else it should be set to `MEMBER_ID_INVALID`. However the spec seems kinda split on if annotations are aggregated: 7.3.2.5 vs Table 8. Maybe this deserves a spec issue to clarify?

Reply:
Annotation is not aggregated based on Figure 4

Reply:
Table 51 says annotations are aggregated though...

Reply:
I thought annotation was out of scope?  We may need to have a design discussion outside of the PR review to figure out where the line gets drawn.

Reply:
Consensus seems to be to treat them as if they were `autoid(SEQUENTIAL)` and file a spec issue.